crates/polars-ops/src/chunked_array/top_k.rs

```diff
@@ -32,46 +41,163 @@ where
+        for arr in ca.downcast_iter() {
+            v.extend(arr.non_null_values_iter());
+        }
+        let values = arg_partition(&mut v, k, descending, TotalOrd::tot_cmp);
+        ChunkedArray::from_slice(ca.name(), values)
+    } else {
+        let mut v = Vec::with_capacity(ca.len());
+        for arr in ca.downcast_iter() {
+            v.extend(arr.iter());
+        }
+        let values = arg_partition(&mut v, k, descending, TotalOrd::tot_cmp);
+        let mut out = ChunkedArray::from_iter(values.iter().copied());
+        out.rename(ca.name());
+        out
+    }
+}
+
 pub fn top_k(s: &[Series], descending: bool) -> PolarsResult<Series> {
-    let src = &s[0];
     let k_s = &s[1];
 
+    polars_ensure!(
+        k_s.len() == 1,
+        ComputeError: "`k` must be a single value for `top_k`."
+    );
+
+    let Some(k) = k_s.cast(&IDX_DTYPE)?.idx()?.get(0) else {
+        polars_bail!(ComputeError: "`k` must be set for `top_k`")
+    };
+
+    let src = &s[0];
+
     if src.is_empty() {
         return Ok(src.clone());
     }
 
-    polars_ensure!(
-        k_s.len() == 1,
-        ComputeError: "k must be a single value."
-    );
+    match src.is_sorted_flag() {
+        polars_core::series::IsSorted::Ascending => {
+            // TopK is the k element in the bottom of ascending sorted array
+            return Ok(src
+                .slice((src.len() - k as usize) as i64, k as usize)
+                .reverse());
+        },
+        polars_core::series::IsSorted::Descending => {
+            return Ok(src.slice(0, k as usize));
+        },
+        _ => {},
+    }
 
-    let k_s = k_s.cast(&IDX_DTYPE)?;
-    let k = k_s.idx()?;
+    let origin_dtype = src.dtype();
 
-    let dtype = src.dtype();
+    let s = src.to_physical_repr();
 
-    if let Some(k) = k.get(0) {
-        let s = src.to_physical_repr();
-        macro_rules! dispatch {
-            ($ca:expr) => {{
-                top_k_impl($ca, k as usize, descending).into_series()
-            }};
-        }
+    match s.dtype() {
+        DataType::Boolean => {
+            Ok(top_k_bool_impl(s.bool().unwrap(), k as usize, descending).into_series())
+        },
+        DataType::String => {
+            top_k_binary_impl(&s.str().unwrap().as_binary(), k as usize, descending)
+                .into_series()
+                .cast(origin_dtype)
+        },
+        DataType::Binary => {
+            Ok(top_k_binary_impl(s.binary().unwrap(), k as usize, descending).into_series())
+        },
+        _dt => {
+            macro_rules! dispatch {
+                ($ca:expr) => {{
+                    top_k_num_impl($ca, k as usize, descending).into_series()
+                }};
+            }
 
-        downcast_as_macro_arg_physical!(&s, dispatch).cast(dtype)
-    } else {
-        Ok(Series::full_null(src.name(), src.len(), dtype))
+            downcast_as_macro_arg_physical!(&s, dispatch).cast(origin_dtype)
```

Comment:
I know this was already using `cast`, but let's `cast_unchecked` here as we know we don't change the type.